jupyter_server_fileid/manager.py

```diff
@@ -37,35 +37,174 @@ def wrapped(self, *args, **kwargs):
+        if not os.path.isabs(proposal["value"]):
+            raise TraitError(f"FileIdManager : {proposal['trait'].name} must be an absolute path")
+        return proposal["value"]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def index(self, path: str) -> Union[int, str, None]:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def get_id(self, path: str) -> Union[int, str, None]:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def get_path(self, id: Union[int, str]) -> Union[int, str, None]:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def move(self, old_path: str, new_path: str) -> Union[int, str, None]:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def copy(self, from_path: str, to_path: str) -> Union[int, str, None]:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def delete(self, path: str) -> None:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def save(self, path: str) -> None:
+        raise NotImplementedError("must be implemented by subclass")
+
+    def get_handlers_by_action(self) -> Dict[str, Optional[Callable[[Dict[str, Any]], Any]]]:
+        """Returns a dictionary whose keys are contents manager event actions
+        and whose values are callables invoked upon receipt of an event of the
+        same action. The callable accepts the body of the event as its only
+        argument. To ignore an event action, set the value to `None`."""
+        raise NotImplementedError("must be implemented by subclass")
+
+
+class ArbitraryFileIdManager(BaseFileIdManager):
+    """
+    File ID manager that works on arbitrary filesystems. Each file is assigned a
+    unique ID. The path is only updated upon calling `move()`, `copy()`, or
+    `delete()`, e.g. upon receipt of contents manager events emitted by Jupyter
+    Server 2.
+    """
+
+    def __init__(self, *args, **kwargs):
+        # pass args and kwargs to parent Configurable
+        super().__init__(*args, **kwargs)
+        # initialize instance attrs
+        self._update_cursor = False
+        # initialize connection with db
+        self.log.info(f"ArbitraryFileIdManager : Configured root dir: {self.root_dir}")
+        self.log.info(f"ArbitraryFileIdManager : Configured database path: {self.db_path}")
+        self.con = sqlite3.connect(self.db_path)
+        self.log.info("ArbitraryFileIdManager : Successfully connected to database file.")
+        self.log.info("ArbitraryFileIdManager : Creating File ID tables and indices.")
+        # do not allow reads to block writes. required when using multiple processes
+        self.con.execute("PRAGMA journal_mode = WAL")
+        self.con.execute(
+            "CREATE TABLE IF NOT EXISTS Files("
+            "id INTEGER PRIMARY KEY AUTOINCREMENT, "
+            "path TEXT NOT NULL UNIQUE"
```

Comment:
I think the `ArbitraryFileIdManager` is going to want to store the `root_dir` value as it can change between invocations and may be necessary.  We should also store _any_ information returned from the Contents events.
```
"root_dir TEXT NOT NULL"
```